### ISSUE TITLE
getindex and setindex! much slower than just accessing array contents directly.

### ISSUE THREAD

**Issue by ali-ramadhan** [CliMA/Oceananigans.jl#13]
Apparently it's 4~5x faster to do operations on `Field.data` instead of `Field` even thought I've inlined `getindex` and `setindex!`, not that it changes things much.
```julia
@inline getindex(f::Field, inds...) = getindex(f.data, inds...)
@inline setindex!(f::Field, v, inds...) = setindex!(f.data, v, inds...)
```
Probably just missing something simple but for now I'll use `Field.data`. Would be nice to figure this out though.

```julia
g = RegularCartesianGrid((100, 100, 100), (10, 10, 10))
f1, f2 = CellField(g), FaceFieldX(g)

function δx1!(g::RegularCartesianGrid, f::CellField, δxf::FaceField)
    for k in 1:g.Nz, j in 1:g.Ny, i in 1:g.Nx
        @inbounds δxf[i, j, k] =  f[i, j, k] - f[decmod1(i, g.Nx), j, k]
    end
end
```
```
julia> @benchmark δx1!(g, f1, f2)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     4.542 ms (0.00% GC)
  median time:      5.007 ms (0.00% GC)
  mean time:        5.120 ms (0.00% GC)
  maximum time:     11.010 ms (0.00% GC)
  --------------
  samples:          975
  evals/sample:     1
```
```julia
function δx2!(g::RegularCartesianGrid, f::CellField, δxf::FaceField)
    for k in 1:g.Nz, j in 1:g.Ny, i in 1:g.Nx
        @inbounds δxf.data[i, j, k] =  f.data[i, j, k] - f.data[decmod1(i, g.Nx), j, k]
    end
end
```
```
julia> @benchmark δx2!(g, f2, f1)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     1.099 ms (0.00% GC)
  median time:      1.198 ms (0.00% GC)
  mean time:        1.253 ms (0.00% GC)
  maximum time:     2.679 ms (0.00% GC)
  --------------
  samples:          3967
  evals/sample:     1
```

**Comment by ali-ramadhan** [comment]
@peterahrens suggested that we might not want to define `setindex!` and `getindex` for fields anyways and instead allow interactions with them through higher levels functions. Worth thinking about, would also make this a non-issue.

But we would still be typing `.data` everywhere for operators.

**Comment by ali-ramadhan** [comment]
> Related to this, I wanted to point something out:
> 
> ```julia
> 
> julia> struct Test
>        a::Array{Float64, 1}
>        end
> 
> julia> import Base: getindex
> 
> julia> getindex(t::Test, inds...) = getindex(t.a, inds...)
> getindex (generic function with 196 methods)
> 
> julia> t = Test(rand(4))
> Test([0.502462, 0.632246, 0.585965, 0.845577])
> 
> julia> @code_lowered t.a[2]
> CodeInfo(
> 731 1 ─ %1 = (Base.arrayref)($(Expr(:boundscheck)), A, i1)                                                                                                  │
>     └──      return %1                                                                                                                                      │
> )
> 
> julia> @code_lowered t[2]
> CodeInfo(
> 1 1 ─ %1 = (Base.getproperty)(t, :a)                                                                                                                        │
>   │   %2 = (Core.tuple)(%1)                                                                                                                                 │
>   │   %3 = (Core._apply)(Main.getindex, %2, inds)                                                                                                           │
>   └──      return %3                                                                                                                                        │
> )
> 
> julia> @code_lowered @inbounds t.a[2]
> CodeInfo(
> 538 1 ─      nothing                                                                                                                                        │
>     │   %2 = (Base.Expr)(:inbounds, true)                                                                                                                   │
>     │   %3 = (Base.esc)(blk)                                                                                                                                │
>     │   %4 = (Base.Expr)(:(=), :val, %3)                                                                                                                    │
>     │   %5 = (Base.Expr)(:local, %4)                                                                                                                        │
>     │   %6 = (Base.Expr)(:inbounds, :pop)                                                                                                                   │
>     │   %7 = (Base.Expr)(:block, %2, %5, %6, :val)                                                                                                          │
>     └──      return %7                                                                                                                                      │
> )
> 
> julia> @code_lowered @inbounds t[2]
> \CodeInfo(
> 538 1 ─      nothing                                                                                                                                        │
>     │   %2 = (Base.Expr)(:inbounds, true)                                                                                                                   │
>     │   %3 = (Base.esc)(blk)                                                                                                                                │
>     │   %4 = (Base.Expr)(:(=), :val, %3)                                                                                                                    │
>     │   %5 = (Base.Expr)(:local, %4)                                                                                                                        │
>     │   %6 = (Base.Expr)(:inbounds, :pop)                                                                                                                   │
>     │   %7 = (Base.Expr)(:block, %2, %5, %6, :val)                                                                                                          │
>     └──      return %7                                                                                                                                      │
> )
> ```
> 
> Note that the lowered code for an *inbounds* `getindex` call is exactly the same for `Test` as it is for it's field, `Test.a`. There is a difference when the array access is not inbounds. So we probably shouldn't see slow down for using `getindex` on `Field` directly if its inbounds.
> 
> _Originally posted by @glwagner in https://github.com/climate-machine/Oceananigans.jl/issues/163#issuecomment-479825567_

**Comment by ali-ramadhan** [comment]
@glwagner Just pasted your comment here where it's more relevant.

That's interesting. I did use `@inbounds` when I tested this in the original post and saw a slowdown. Seems like it's worth investigating again, especially with `@code_lowered`, Cthulu.jl, etc. as you're showing that they should compile to identical code!

**Comment by vchuravy** [comment]
You are probably missing a `@propagate_inbounds`

**Comment by glwagner** [comment]
Should we close this?

**Comment by ali-ramadhan** [comment]
It looks like this issue is actually resolved now (at least as of Julia 1.5.1 + Oceananigans v0.38.0)!

```julia
using BenchmarkTools, Oceananigans
grid = RegularCartesianGrid(size=(256, 256, 256), extent=(1, 1, 1))
model = IncompressibleModel(grid=grid)
u, v, w = model.velocities

function δx_field!(δf, f, grid)
    for k in 1:grid.Nz, j in 1:grid.Ny, i in 1:grid.Nx
        @inbounds δf[i, j, k] =  f[i, j, k] - f[i-1, j, k]
    end
end

δx_field!(u, v, grid)
@benchmark δx_field!(u, v, grid)
```

```
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     15.779 ms (0.00% GC)
  median time:      17.309 ms (0.00% GC)
  mean time:        18.481 ms (0.00% GC)
  maximum time:     34.617 ms (0.00% GC)
  --------------
  samples:          271
  evals/sample:     1
```

```julia
function δx_data!(δf, f, grid)
    for k in 1:grid.Nz, j in 1:grid.Ny, i in 1:grid.Nx
        @inbounds δf.data[i, j, k] =  f.data[i, j, k] - f.data[i-1, j, k]
    end
end

δx_data!(u, v, grid)

@benchmark δx_data!(u, v, grid)
```

```
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     15.618 ms (0.00% GC)
  median time:      16.714 ms (0.00% GC)
  mean time:        18.068 ms (0.00% GC)
  maximum time:     32.793 ms (0.00% GC)
  --------------
  samples:          277
  evals/sample:     1
```
